src/lib.rs

```diff
@@ -23,389 +32,524 @@ unsafe impl Send for LDAP {}
+    fn ldap_search_ext_s(ldap: *mut LDAP, base: *const c_char, scope: c_int,
+                         filter: *const c_char, attrs: *const *const c_char,
+                         attrsonly: c_int, serverctrls: *mut *mut LDAPControl,
+                         clientctrls: *mut *mut LDAPControl, timeout: *mut timeval,
+                         sizelimit: c_int, res: *mut *mut LDAPMessage) -> c_int;
+    fn ldap_unbind_ext_s(ldap: *mut LDAP, sctrls: *mut *mut LDAPControl, cctrls: *mut *mut LDAPControl) -> c_int;
 }
 
+/// A typedef for an LDAPResponse type.
+///
+/// LDAP responses are organized as vectors of mached entities. Typically, each entity is
+/// represented as a map of attributes to list of values.
+///
+pub type LDAPResponse = Vec<HashMap<String,Vec<String>>>;
+
+
+/// A high level abstraction over the raw OpenLDAP functions.
+///
+/// A `RustLDAP` object hides raw OpenLDAP complexities and exposes a simple object that is
+/// created, configured, and queried. Methods that call underlying OpenLDAP calls that can fail
+/// will raise an `errors::LDAPError` with additional details.
+///
+/// Using a `RustLDAP` object is easy!
+///
 pub struct RustLDAP {
-	// Have the raw pointer to it so we can pass it into internal functions
-	ldap_ptr: *mut LDAP
+    /// A pointer to the underlying OpenLDAP object.
+    ldap_ptr: *mut LDAP,
 }
 
 impl Drop for RustLDAP {
+    fn drop(&mut self){
+        // Unbind the LDAP connection, making the C library free the LDAP*.
+        let rc = unsafe { ldap_unbind_ext_s(self.ldap_ptr, ptr::null_mut(), ptr::null_mut()) };
+
+        // Make sure it actually happened.
+        if rc != codes::results::LDAP_SUCCESS {
+            unsafe {
+                // Hopefully this never happens.
+                let raw_estr = ldap_err2string(rc as c_int);
+                panic!(CStr::from_ptr(raw_estr).to_owned().into_string().unwrap());
+            }
+        }
+    }
+}
 
-	fn drop(&mut self){
-
-		//unbind the LDAP connection, making the C library free the LDAP*
-		let rc = unsafe { ldap_unbind_ext_s(self.ldap_ptr, ptr::null_mut(), ptr::null_mut()) };
+/// A trait for types that can be passed as LDAP option values.
+///
+/// Underlying OpenLDAP implementation calls for option values to be passed in as *const c_void,
+/// while allowing values to be i32 or string. Using traits, we implement function overloading to
+/// handle i32 and string option value types.
+///
+pub trait LDAPOptionValue {
+    fn as_cvoid_ptr(&self) -> *const c_void;
+}
 
-		//make sure it actually happened
-		if rc != codes::results::LDAP_SUCCESS {
-			unsafe { //hopefully this never happens
-				let raw_estr = ldap_err2string(rc as c_int);
-				panic!(CStr::from_ptr(raw_estr).to_owned().into_string().unwrap());
-			}
+impl LDAPOptionValue for str {
+    fn as_cvoid_ptr(&self) -> *const c_void {
+        let string = CString::new(self).unwrap();
+        unsafe {
+            return mem::transmute(string.as_ptr());
```

Comment:
As clever as this code is, I believe this is actually memory-unsafe.

`string` is only in scope from line 112 until line 116. `string.as_ptr()` will simply obtain the memory address of the storage used by `string`. Transmuting it will not copy that data anywhere either. So, the void* being returned is actually a dangling pointer, since the memory for `string` *will* be deallocated on line 116.